File: algorithms/algorithms-cplusplus/leetcode/Question_0313_Super_Ugly_Number.cc
#include <limits.h>
#include <vector>

using namespace std;

class Solution {
 public:
  int nthSuperUglyNumber(int n, vector<int>& primes) {
    vector<long long> dp(n, INT_MAX);
    dp[0] = 1;
    vector<int> index(primes.size(), 0);
    for (int i = 1; i < n; i++) {
      for (int j = 0; j < index.size(); j++) {
        dp[i] = min(dp[i], primes[j] * dp[index[j]]);
      }
      for (int j = 0; j < index.size(); j++) {
        if (dp[i] == primes[j] * dp[index[j]]) {
          index[j]++;
        }
      }
    }
    return (int)dp[n - 1];
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0313_Super_Ugly_Number.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Arrays;

public class Question_0313_Super_Ugly_Number {
    public int nthSuperUglyNumber(int n, int[] primes) {
        long[] dp = new long[n];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 1;
        int[] indexes = new int[primes.length];
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < primes.length; j++) {
                dp[i] = Math.min(dp[i], dp[indexes[j]] * primes[j]);
            }
            for (int j = 0; j < primes.length; j++) {
                if (dp[i] == dp[indexes[j]] * primes[j]) {
                    indexes[j]++;
                }
            }
        }
        return (int)dp[n - 1];
    }
}
